libpimeval/src/pimCmd.h

```diff
@@ -380,30 +385,32 @@ class pimCmdFunc2 : public pimCmd
   }
 };
 
-//! @class  pimCmdedSum
-//! @brief  Pim CMD: RedSum non-ranged/ranged
-template <typename T> class pimCmdRedSum : public pimCmd
+//! @class  pimCmdReductiom
+//! @brief  Pim CMD: Reduction non-ranged/ranged
+template <typename T>
+class pimCmdReduction : public pimCmd
 {
 public:
-  pimCmdRedSum(PimCmdEnum cmdType, PimObjId src, T* result)
+  pimCmdReduction(PimCmdEnum cmdType, PimObjId src, void* result)
     : pimCmd(cmdType), m_src(src), m_result(result)
   {
-    assert(cmdType == PimCmdEnum::REDSUM);
+    assert(cmdType == PimCmdEnum::REDSUM || cmdType == PimCmdEnum::REDMIN || cmdType == PimCmdEnum::REDMAX);
   }
-  pimCmdRedSum(PimCmdEnum cmdType, PimObjId src, T* result, uint64_t idxBegin, uint64_t idxEnd)
-    : pimCmd(cmdType), m_src(src), m_result(result), m_idxBegin(idxBegin), m_idxEnd(idxEnd)
+  pimCmdReduction(PimCmdEnum cmdType, PimObjId src, void* result, uint64_t idxBegin, uint64_t idxEnd)
+    : pimCmd(cmdType), m_src(src), m_result(result), m_idxBegin(idxBegin)
   {
-    assert(cmdType == PimCmdEnum::REDSUM_RANGE);
+    assert(cmdType == PimCmdEnum::REDSUM || cmdType == PimCmdEnum::REDMIN || cmdType == PimCmdEnum::REDMAX || cmdType == PimCmdEnum::REDSUM_RANGE || cmdType == PimCmdEnum::REDMIN_RANGE || cmdType == PimCmdEnum::REDMAX_RANGE);
+    if (idxEnd) m_idxEnd = idxEnd;
   }
-  virtual ~pimCmdRedSum() {}
+  virtual ~pimCmdReduction() {}
   virtual bool execute() override;
   virtual bool sanityCheck() const override;
   virtual bool computeRegion(unsigned index) override;
   virtual bool updateStats() const override;
 protected:
   PimObjId m_src;
-  T* m_result;
-  std::vector<T> m_regionSum;
+  void* m_result;
+  std::vector<T> m_regionResult;
```

Comment:
Using data type T for region partial result does not look correct to me here. When T is int/uint 8/16/32, we need to store 64-bit partial sum results.

Reply:
for reduction sum PIMSim calls this with either int64 or uint64.

Reply:
I see. Then it's good